README
Dependencies-Searcher

Dependencies::Searcher - Search recursively dependencies used in a Perl module's root directory and build a report that can be used as a Carton cpanfile.

Maybe you don't want to have to list all the dependencies of your Perl application by hand and want an automated way to build it. Maybe you forgot to do it for a long time ago. During this time, you've add lots of CPAN modules. Carton is here to help you manage dependencies between your development environment and production, but how to keep track of the list of modules you will pass to to Carton?

Event if it is a no brainer to keep track of this list, it can be much better not to have to do it.

You will need a tool that will check for any 'requires' or 'use' in your module package, and report it into a file that could be used as a Carton cpanfile. Any duplicated entry will be removed and modules versions will be checked and made available. Core modules will be ommited because you don't need to install them.

This project has begun because it happens to me, and I don't want to search for modules to install by hand, I just want to run a simple script that update the list in a simple way. It was much more longer to write the module than to search by hand but I wish it will be usefull.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Dependencies::Searcher

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dependencies-Searcher

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Dependencies-Searcher

    CPAN Ratings
        http://cpanratings.perl.org/d/Dependencies-Searcher

    Search CPAN
        http://search.cpan.org/dist/Dependencies-Searcher/


LICENSE AND COPYRIGHT

Copyright (C) 2013 smonff

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.



Changes
Revision history for Dependencies-Searcher

0.063 2014-02-16 19:29
 [FIX]
 - Fix #48 : some tests fails because we use modules that where supposed
   to not be in core but are in recent version of Perl (especially with 
   Module::CoreList)
 - If we don't find a specific module version, we format the cpanfile
   line in the right way anyway

0.062 2014-02-16 00:09
 [FIX]
 - Now use the standard cpanfile syntax
 - Now use appropriate version numbers (thanx chorny) 

0.06.1 2014-01-06 09:25
 [FIX]
 - A pod error, need to deploy it again to run the tests

0.06 2014-01-05 21:28
 [FIX]
 - First stable release !
 - Only change some typos, since all the issues have been resolved

0.05_11 2013-11-10 20:55
 [EHANCEMENT]
 - Remove some tests that weren't possible to pass

0.05_10 2013-11-09 23:44
 [FIX]
 - Use a full path for Ack use tests, because the simple 'ack' command is not portable
 - Fix bug #39, modules array that was supposed to be in scalar context but wasn't
 - Broken link in POD
 [EHANCEMENT]
 - Moved some tests from the full-impl test suite to ack-requester test suite

0.05_09 2013-11-03 23:08
 [FIX]
 - Fix the issue #37, POD lines starting by use or requires are not considered as code anymore
 - Fix the issue #30 , using Path::Class for making the module portable (especially path separators) 
 - A pod error
 [EHANCEMENTS]
 - Tests for found patterns (POD or code)
 - Activate tests for POD
 - Tests for cpanfile

0.05_08 2013-10-20 21:58
 [EHANCEMENTS]
 - Format and clean code
 - Write AckRequester Documentation
 - Complete and correct Dependencies::Searcher documentation

0.05_07 2013-10-04 21:23
 [FIX]
 - Close bug #25 about modules in corelist with a fresher version 
 - New regex for cleaning, now everything is clean
 - Module '1' were appearing when merging 2 use and require arrays
 - Force the use of IPC::Run and add it (and IPC::Cmd) to Makefile.PL requirements
 [EHANCEMENTS]
 - Better logs 
 - Make log file directory cross-platform (thanx mishin)

0.05_06 2013-09-24 08:57
 [FIX]
 - Translate some french doc and remove some non UTF-8 stuff

0.05_05 2013-09-23 20:45
 [EHANCEMENTS]
 - Log to a /tmp file
 - Removes version numbers
 [FIX]
 - Stupid typo in Makefile.PL prerequisites

0.05_04 2013-09-22 20:02
 [EHANCEMENTS]
 - Upgrade documentation
 - Logs system with File::Stamped & Log::Minimal
 - Use Module::Version's Perl API, not command line (avoid problems reported by Smoke testers)
 - Now implement the Module::CoreList 2.99 instead of 2.4902  to use it's is_core() method (avoids to use another IPC::Cmd)  

0.05_03 2013-09-17 07:51
 [FIXES]
 - A pod error
 - Add App::Ack and Module::Version to TEST_REQUIRES in Makefile.PL

0.05_02 2013-09-16 22:11
 [FIXES]
 - get_modules() method rewriting, using IPC::Cmd, not exec() or system()
 - Fix crash when not finding any use or require
 - Make the module more portable by using Cwd to get some dynamic path and avoid relative path problems
 [EHANCEMENTS]
 - Make Dependencies::Searcher using Dependencies::Searcher::AckRequester
 - Build an independant module for the Ack part, that returns an array containing interesting lines
 - Add a MIN_PERL_VERSION (5.010) in Makefile.PL as recommanded at CPANTS

0.05_01 2013-09-10 23:10
 [EHANCEMENTS]
 - Test Ack availability in tests
 - Remove tests that asks for a static number of results (non-predictable) 

0.04    2013-09-09 17:38:24 
 [FIXES]
 - [RT #88528] Dependencies::Searcher::Utils have been removed
   trying to clean up test failures

0.03    2013-09-04 01:45
 [FIXES]
 - [RT #88392] Fix a bug about version number in Changes file
 - Use the CPAN App::Ack, not the the one from packages managers 

 [EHANCEMENTS]
 - Can now get the count of core / non-core modules 
 - Add some Moose helper traits for ArrayRef attributes, now you can
   add and count and the code is easiest to read !
 - Clean a lot of logs
 - Change the dissociate() implementation with Moose traits 
 - Add clean_everything(), uniq() and dissociate() methods
 - Add two Moose attributes for core / non-core modules
 - Ignore use warnings FATAL => 'all'; 
 - Ignore parsed modules minimal perl versions.
 - Moosification of the module.

0.02    2013-09-02 20:48
 - First version, released on an unsuspecting world. 
 - Exile the module from the Shurf::Wax distribution because it seems 
   it could be re-used for other projects and by you. 
      
0.01    2013-07-21
 - Begin it as a script embedded in a Catalyst app.
   See https://github.com/smonff/Shurf/commits/master/script/uses_finder.pl
        



lib/Dependencies/Searcher.pm
package Dependencies::Searcher;

use 5.010;
use Data::Printer;
use feature qw(say);
# Since 2.99 a is_core() method is available :)
use Module::CoreList 2.99;
use Module::Version 'get_version';
use autodie;
use Moose;
use IPC::Cmd qw[can_run run];
use Dependencies::Searcher::AckRequester;
use Log::Minimal env_debug => 'LM_DEBUG';
use File::Stamped;
use IO::File;
use File::HomeDir;
use File::Spec::Functions qw(catdir catfile);
use Version::Compare;
use Path::Class;
use ExtUtils::Installed;

our $VERSION = '0.063';

=head1 NAME

Dependencies::Searcher - Search recursively dependencies used in a
module's directory and build a report that can be used as a L<Carton|Carton>
cpanfile.

=cut

=head1 SYNOPSIS

    use Dependencies::Searcher;

    my $searcher = Dependencies::Searcher->new();
    my @elements = $searcher->get_files();
    my @uses = $searcher->get_modules($path, "use");
    my @uniq_modules = $searcher->uniq(@uses);

    $searcher->dissociate(@uniq_modules);

    $searcher->generate_report($searcher->non_core_modules);

    # Prints
    # requires Data::Printer, 0.35
    # requires Moose, 2.0602
    # requires IPC::Cmd
    # requires Module::Version
    # ...

=cut

=head1 DESCRIPTION

Maybe you don't want to have to list all the dependencies of your Perl
application by hand and want an automated way to build it. Maybe you
forgot to do it for a long time ago. Or just during a short period.
Anyway, you've add lots of CPAN modules. L<Carton|Carton> is here to help you
manage dependencies between your development environment and
production, but how to keep track of the list of modules you will pass
to L<Carton|Carton>?

Event if it is a no brainer to keep track of this list by adding it by
hand, it can be much better not to have to do it.

You will need a tool that will check for any I<requires> or I<use> in
your module package, and report it into a file that could be used as an
input L<Carton|Carton> cpanfile. Any duplicated entry will be removed and
modules versions will be checked and made available. Core modules will be
ommited because you don't need to install them (except in some special
case, see C<dissociate()> documentation).

This project has begun because it has happened to me, and I don't want
to search for modules to install by hand, I just want to run a simple
script that update the list in a convenient way. It was much more
longer to write the module than to search by hand so I wish it could
be useful for you now.

This module is made to search dependencies for I<installed
distributions>, it is not supposed to manage anything else.

=cut

=head1 WHY ISN'T IT JUST ANOTHER MODULE::SCANDEPS ?

Module::ScanDeps is a bi-dimentional recursive scanner: it features
dependencies and directories recursivity.

Dependencies::Searcher only found direct dependencies, not
dependencies of dependencies, it scans recursively directories but not
dependencies..

These direct dependencies are passed to the Perl toolchain (cpanminus)
that will take care of any recursive dependencies.

=cut

# Init parameters
has 'non_core_modules' => (
    traits     => ['Array'],
    is         => 'rw',
    isa        => 'ArrayRef[Str]',
    default    => sub { [] },
    handles    => {
	add_non_core_module    => 'push',
	count_non_core_modules => 'count',
    },
);

has 'core_modules' => (
    traits     => ['Array'],
    is         => 'rw',
    isa        => 'ArrayRef[Str]',
    default => sub { [] },
    handles    => {
	add_core_module    => 'push',
	count_core_modules => 'count',
    },
);

# Log stuff here
local $ENV{LM_DEBUG} = 0; # 1 for debug logs, 0 for info

my $work_path = File::HomeDir->my_data;
my $log_fh = File::Stamped->new(
    pattern => catdir($work_path,  "dependencies-searcher.log.%Y-%m-%d.out"),
);

say("tail -vf $work_path for log");

# Overrides Log::Minimal PRINT
$Log::Minimal::PRINT = sub {
    my ( $time, $type, $message, $trace) = @_;
    print {$log_fh} "$time [$type] $message\n";
};

infof('  * * * * * * * * * * * * * * * * * * * *');
infof('* H E R E   I S   A   N E W   S E A R C H *');
infof('  * * * * * * * * * * * * * * * * * * * *');

infof("Dependencies::Searcher $VERSION debugger init.");
infof("Log file available in " . $work_path);
# End of log init

sub get_modules {
    # @path contains files and directories
    my ($self, $pattern, @path) = @_;

    debugf("Ack pattern : " . $pattern);

    # The regex add the terminal semicolon at the end of the line to
    # make the difference between comments and code, because "use" is
    # a word that you can find often in a POD section, more much in
    # the beginning of line than you could think
    $pattern = "$pattern" . qr/.+;$/;

    my @params = ('--perl', '-hi', $pattern, @path);
    foreach my $param (@params) {
	debugf("Param : " . $param);
    }

    my $requester = Dependencies::Searcher::AckRequester->new();

    my $ack_path = $requester->get_path();
    debugf("Ack path : " . $ack_path);

    my $cmd_use = $requester->build_cmd(@params);

    my @moduls = $requester->ack($cmd_use);
    infof("Found $pattern modules : " . @moduls);

    if ( defined $moduls[0]) {
	if ($moduls[0] =~ m/^use/ or $moduls[0] =~ m/^require/) {
	    return @moduls;
	} else {
	    critf("Failed to retrieve modules with Ack");
	    die "Failed to retrieve modules with Ack";
	}
    } else {
	say "No use or require found !";
    }
}

sub get_files {
    my $self = shift;
    # Path::Class  functions allows a more portable module
    my $lib_dir = dir('lib');
    my $make_file = file('Makefile.PL');
    my $script_dir = dir('script');

    my @structure;
    $structure[0] = "";
    $structure[1] = "";
    $structure[2] = "";
    if (-d $lib_dir) {

	$structure[0] = $lib_dir;

    } else {
	# TODO : TEST IF THE PATH IS OK ???
	die "Don't look like we are working on a Perl module";
    }

    if (-f $make_file) {
	$structure[1] = $make_file;
    }

    if (-d $script_dir) {
	$structure[2] = $script_dir;
    }

    return @structure;
}

# Generate a "1" when merging if one of both is empty
# Will be clean in make_it_real method
sub merge_dependencies {
    my ($self, @uses, @requires) = @_;
    my @merged_dependencies = (@uses, @requires);
    infof("Merged use and require dependencies");
    return @merged_dependencies;
}

# Remove special cases that aren't need at all
sub make_it_real {
    my ($self, @merged) = @_;
    my @real_modules;
    foreach my $module ( @merged ) {
	push(@real_modules, $module) unless

	$module =~ m/say/

	# Describes a minimal Perl version
	or $module =~ m/^use\s[0-9]\.[0-9]+?/
	or $module =~ m/^use\sautodie?/
	or $module =~ m/^use\swarnings/
	# Kind of bug generated by merge_dependencies() when there is
	# only one array to merge
	or $module =~ m/^1$/
	or $module =~ m/^use\sDependencies::Searcher/;
    }
    return @real_modules;
}

# Clean correct lines that can't be removed
sub clean_everything {
    my ($self, @dirty_modules) = @_;
    my @clean_modules = ();

    foreach my $module ( @dirty_modules ) {

	debugf("Dirty module : " . $module);

	# remove the 'use' and the space next
	$module =~ s{
			use \s
		}
		    {}xi; # Empty subtitution

	# remove the require, quotes and the space next
	# but returns the captured module name (non-greedy)
	# i = not case-sensitive
	$module =~ s{
			requires \s
			'
			(.*?)
			'
		}{$1}xi; # Note -> don't insert spaces here

	# Remove the ';' at the end of the line
	$module =~ s/ ; //xi;

	# Remove any qw(xxxxx xxxxx) or qw[xxx xxxxx]
	# '\(' are for real 'qw()' parenthesis not for grouping
	# Also removes empty qw()
	$module =~ s{
			\s qw
			\(
			([A-Za-z]+(\s*[A-Za-z]*))*
			\)
		}{}xi;
	$module =~ s{
			\s qw
			\[
			([A-Za-z]+(_[A-Za-z]+)*(\s*[A-Za-z]*))*
			\]
		}
		    {}xi; # Empty subtitution

	# Remove method names between quotes (those that can be used
	# without class instantiation)
	$module =~ s{
			\s
			'
			[A-Za-z]+(_[A-Za-z]+)*
			'
		}
		    {}xi; # Empty subtitution

	# Remove dirty bases and quotes.
	# This regex that substitute My::Module::Name
	# to a "base 'My::Module::Name'" by capturing
	# the name in a non-greedy way
	$module =~ s{
			base \s
			'
			(.*?)
			'
		}
		    {$1}xi;

	# Remove some warning sugar
	$module =~ s{
			([a-z]+)
			\s FATAL
			\s =>
			\s 'all'
		}
		    {$1}xi;

	# Remove version numbers
	# See "a-regex-for-version-number-parsing" :
	# http://stackoverflow.com/questions/82064/
	$module =~ s{
			\s
			(\*|\d+(\.\d+)
			    {0,2}
			    (\.\*)?)$
		}
		    {}x;

	# Remove configuration stuff like env_debug => 'LM_DEBUG' but
	# the quoted words have been removed before
	$module =~ s{
			\s
			([A-Za-z]+(_[A-Za-z]+)*( \s*[A-Za-z]*))*
			\s
			=>
		}
		    {}xi;

	debugf("Clean module : " . $module);
	push @clean_modules, $module;
    }
    return @clean_modules;
}


sub uniq {
    my ($self, @many_modules) = @_;
    my @unique_modules = ();
    my %seen = ();
    foreach my $element ( @many_modules ) {
	next if $seen{ $element }++;
	debugf("Uniq element added : " . $element);
	push @unique_modules, $element;
    }
    return @unique_modules;
}

sub dissociate {
    my ($self, @common_modules) = @_;

    foreach my $nc_module (@common_modules) {

	# The old way before 2.99 corelist
	# my $core_list_answer = `corelist $nc_module`;

	my $core_list_answer = Module::CoreList::is_core($nc_module);

	if (
	    # "$]" is Perl version
	    (exists $Module::CoreList::version{ $] }{"$nc_module"})
	    or
	    # In case module don't have a version number
	    ($core_list_answer == 1)
	) {

	    # A module can be in core but the wanted version can be
	    # more fresh than the core one...
	    # Return the most recent version
	    my $mversion_version = get_version($nc_module);
	    # Return the corelist version
	    my $corelist_version = $Module::CoreList::version{ $] }{"$nc_module"};

	    debugf("Mversion version : " . $mversion_version);
	    debugf("Corelist version : " . $corelist_version);

	    # Version::Compare warns about versions numbers with '_'
	    # are 'non-numeric values'
	    $corelist_version =~ s/_/./;
	    $mversion_version =~ s/_/./;

	    # It's a fix for this bug
	    # https://github.com/smonff/dependencies-searcher/issues/25
	    # Recent versions of corelist modules are not include in
	    # all Perl versions corelist
	    if (&Version::Compare::version_compare(
		$mversion_version, $corelist_version
	    ) == 1) {
		infof(
		    $nc_module . " version " . $mversion_version .
		    " is in use but  " .
		    $corelist_version .
		    " is in core list"
		);
		$self->add_non_core_module($nc_module);
		infof(
		    $nc_module .
		    " is in core but has been added to non core " .
		    "because it's a fresh core"
		);
		next;
	    }

	    # Add to core_module

	    # The old way
	    # You have to push to an array ref (Moose)
	    # http://www.perlmonks.org/?node_id=695034
	    # push @{ $self->core_modules }, $nc_module;

	    # The "Moose" trait way
	    # http://metacpan.org/module/Moose::Meta::Attribute::Native::Trait::Array
	    $self->add_core_module($nc_module);
	    infof($nc_module . " is core");

	} else {
	    $self->add_non_core_module($nc_module);
	    infof($nc_module . " is not in core");
	    # push @{ $self->non_core_modules }, $nc_module;
	}
    }
}

# Open a file handle to > cpanfile
sub generate_report {

    my $self = shift;

    #
    # TODO !!! Check if the module is installed already with
    # ExtUtils::Installed. If it it not, cry that
    # Dependencies::Searcher is designed to be used in the complete env
    #

    open my $cpanfile_fh, '>', 'cpanfile' or die "Can't open cpanfile : $:!";

    foreach my $module_name ( @{$self->non_core_modules} ) {

	my $version = get_version($module_name);

	# if not undef
	if ($version) {
	    debugf("Module + version : " . $module_name . " " . $version);

	    # Add the "requires $module_name\n" to the next line of the file
	    chomp($module_name, $version);

	    if ($version =~ m/[0-9]\.[0-9]+/ ) {
		say $cpanfile_fh "requires '$module_name', '$version';";
	    } # else : other case ?

	} else {
	    debugf("Module + version : " . $module_name);
	    say $cpanfile_fh "requires '$module_name';";
	}

    }

    close $cpanfile_fh;
    infof("File has been generated and is waiting for you");
}

1;

__END__

=pod

=head1 SUBROUTINES/METHODS

=head2 get_files()

C<get_files()> returns an array containing which file or directories has
been found in the current root distribution directory. We suppose it
can find dependancies in 3 different places :

=over 2

=item * files in C<lib/> directory, recursively

=item * C<Makefile.PL>

=item * C<script/> directory, i.e. if we use a Catalyst application

=item * maybe it should look in C<t/> directory (todo)

=back

If the C<lib/> directory don't exist, the program die because we
consider we are not into a plain old Perl Module.

This is work in progress, if you know other places where we can find
stuff, please report a bug.

=cut

=head2 get_modules("pattern", @elements)

You must pass a pattern to search for, and the elements (files or
directories) where you want to search (array of strings from C<get_files()>).

These patterns should be C<^use> or C<^require>.

Then, Ack will be used to retrieve modules names into lines containing
patterns and return them into an array (containing also some dirt).
See L<Dependencies::Searcher::AckRequester> for more informations.

=cut

=head2 merge_dependencies(@modules, @modules)

Simple helper method that will merge C<use> and C<require> arrays if you
search for both. Return an uniq array. It got a little caveat, see
CAVEATS.

=cut

=head2 make_it_real(@modules)

Move dependencies lines from an array to an another unless it is
considered as a special case : minimal Perl versions, C<use autodie>,
C<use warnings>. These stuff has to be B<removed>. Return a I<real
modules> array (I<real interresting> modules).

=cut

=head2 clean_everything(@modules)

After removing irrelevant stuff, we need to B<clean> what is leaving
and is considered as being crap (not strictly <CName::Of::Module>) but
needs some cleaning. We are going to remove everything but the module
name (even version numbers).

This code section is well commented (because it is regex-based) so,
please refer to it directly.

It returns an array of I<clean modules>.

=cut

=head2 uniq(@modules)

Make each array element uniq, because one dependency can be found many
times. Return an array of unique modules.

=cut

=head2 dissociate(@modules)

Dissociate I<core> / I<non-core> modules using the awesome
C<Module::Corelist::is_core method>, that search in the current Perl
version if the module is from Perl core or not. Note that results can
be different according to the environment.

More, B<you can have two versions of the same module installed on your
environment> (even if you use L<local::lib|local::lib> when you
install a recent version of a file that has been integrated into Perl
core (this version hasn't necessary been merged into core).

So C<dissociate()> checks both and compares it, to be sure that the found core
module is the "integrated" version, not a fresh one that you have
installed yourself. If it is fresh, the module is considered as a I<non-core>.

This method don't return anything, but it stores found dependencies on the two
C<core_modules> and C<non_core_modules> L<Moose|Moose> attributes arrays.

=cut

=head2 generate_report()

Generate the C<cpanfile> for L<Carton|Carton>, based on data contained into
C<core_modules> and C<non_core_modules> attributes, with optionnal
version number (if version number can't be found, dependency name is
print alone).

Generate a hash containing the modules could be achieved. Someday.

=cut

=head2 Log::Minimal::PRINT override

Just override the way Log::Minimal is used. See LOGGING AND DEBUGGING
for more informations.

=cut

=head1 LOGGING AND DEBUGGING

This module has a very convenient logging system that use
L<Log::Minimal|Log::Minimal> and L<File::Stamped|File::Stamped> to
write to a file that you will find in the directory where local
applications should store their internal data for the current
user. This is totally portable (Thanks to Nikolay Mishin
(mishin)). For exemple, on a Debian-like OS :

    ~/.local/share/dependencies-searcher.[y-M-d].out

To debug and use these logs :

    $ tail -vf ~/local/share/dependencies-searcher.[y-M-d].out

For more information on how to configure log level, read
L<Log::Minimal|Log::Minimal> documentation.

For a simple exemple on how to use it, see this blog post http://bit.ly/1lJwyX7

=head1 CAVEATS

=head2 Low Win32 / Cygwin support

This module was'nt supposed to run under Win32 / Cygwin environments
because it was using non portable code with slashes. I hope this gets
better since it has been rewritten using L<Path::Class|Path::Class>
but it still need some testing.

It also us-e Ack as a hack through a system command even if it was not
supposed to be used like that. Yes, this is dirty. Yes, I plan to change
things, even if Ack do the stuff proudly this way.

Thanks to cpantesters.org community reports, things should go better and
better.

=cut

=head1 BUGS

Please report any bugs or feature requests to
C<bug-dependencies-searcher at  rt.cpan.org>, or through the web
interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dependencies-Searcher>.
I will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 TODOs

Most of the time, todos and features are on Github and Questub.
See https://github.com/smonff/dependencies-searcher/issues

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dependencies::Searcher

You can also look for information at:

    See https://github.com/smonff/dependencies-searcher/

=over 2

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dependencies-Searcher>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Dependencies-Searcher>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Dependencies-Searcher>

=item * Search CPAN

L<http://search.cpan.org/dist/Dependencies-Searcher/>

=back

=head1 AUTHOR

smonff, C<< <smonff at gmail.com> >>

=head1 CONTRIBUTORS

=over

=item * Nikolay Mishin (mishin) helps to make it more cross-platform

=item * Alexandr Ciornii (chorny) advises on version numbers

=back

=cut

=head1 ACKNOWLEDGEMENTS

=over

=item * Brian D. Foy's L<Module::Extract::Use|Module::Extract::Use>

Was the main inspiration for this one. First, I want to use it for my needs
but it was not recursive...

See L<https://metacpan.org/module/Module::Extract::Use>

=item * L<Module::CoreList|Module::CoreList>

What modules shipped with versions of perl. I use it extensively to detect
if the module is from Perl Core or not.

See L<http://perldoc.perl.org/Module/CoreList.html>

=item * Andy Lester's Ack

I've use it as the main source for the module. It was pure Perl so I've choose
it, even if Ack is not meant for being used programatically, this use do the
job.

See L<http://beyondgrep.com/>

=back

See also :

=over 2

=item * https://metacpan.org/module/Perl::PrereqScanner

=item * http://stackoverflow.com/questions/17771725/

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2013 smonff.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut




lib/Dependencies/Searcher/Utils.pm
package Dependencies::Searcher::Utils;

use 5.006;
use strict;
use warnings FATAL => 'all';
use Data::Printer;
use feature qw(say);
use Module::CoreList qw();
use autodie;
use Moose;

=head1 NAME

Dependencies::Searcher::Utils - Provides main treatments for Dependancies::Searcher

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';


=head1 SYNOPSIS

Perhaps a little code snippet.

    use Dependencies::Searcher;

    my $foo = Dependencies::Searcher->new();
    ...


=head1 SUBROUTINES/METHODS

=head2 get_modules()

=cut

=head2 get_files()

=cut

sub get_modules {
    my $self = shift;
    # say $self;
    # p @_;
    my ($params, $patrn, $p) = @_;
    my $request = "$params $patrn $p";
    # p $request;
    my @moduls = `ack $request`;


    if ( defined $moduls[0]) {
	if ($moduls[0] =~ m/^use/ or $moduls[0] =~ m/^require/) {
	    return @moduls;
	} else {
	    die "Failed to retrieve modules with Ack";
	}
    } else {
	say "No $patrn found !";
    }
}

sub get_files {
    my $self = shift;
    my @structure;
    $structure[0] = "";
    $structure[1] = "";
    $structure[2] = "";
    if (-d "lib") {
	$structure[0] = "lib";
    } else {
	die "Don't look like we are working on a Perl module";
    }

    if (-f "Makefile.PL") {
	$structure[1] = "Makefile.PL";
    }

    if (-d "script") {
	$structure[2] = "script";
    }
    # p @structure;
    return @structure;
}

=head1 AUTHOR

smonff, C<< <smonff at gmail.com> >>

=head1 BUGS

See dstribution main module

Bugtracker : https://github.com/smonff/dependencies-searcher/issues

Please report any bugs or feature requests to C<bug-dependencies-searcher at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dependencies-Searcher>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dependencies::Searcher

=head1 LICENSE AND COPYRIGHT

Copyright 2013 smonff.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Dependencies::Searcher


lib/Dependencies/Searcher/AckRequester.pm
package Dependencies::Searcher::AckRequester;

use 5.010;
use Data::Printer;
use feature qw(say);
use Module::CoreList qw();
use autodie;
use Moose;
use IPC::Cmd qw[can_run run];
use IPC::Run;
use Log::Minimal env_debug => 'LM_DEBUG';
use File::Stamped;
use File::HomeDir;
use File::Spec::Functions qw(catdir catfile);


# These modules will be used throught a system call
# App::Ack;

has 'full_path' => (
  is  => 'rw',
  isa => 'Str',
);

$IPC::Cmd::USE_IPC_RUN = 1;

local $ENV{LM_DEBUG} = 1; # 1 for debug logs, 0 for info

my $work_path = File::HomeDir->my_data;
my $log_fh = File::Stamped->new(
    pattern => catdir($work_path,  "dependencies-searcher.log.%Y-%m-%d.out"),
);

# Overrides Log::Minimal PRINT
$Log::Minimal::PRINT = sub {
    my ( $time, $type, $message, $trace) = @_;
    print {$log_fh} "$time [$type] $message\n";
};

sub get_path {
    my $self = shift;

    my $tmp_full_path = can_run('ack');

    if ($tmp_full_path) {
	$self->full_path($tmp_full_path);
	debugf("Ack full path : " . $self->full_path);
	return $self->full_path;
    } else {
	critf('Something goes wrong with Ack path or IPC::Run is not available !');
    }
}

sub build_cmd {

    my ($self, @params) = @_;

    my @cmd = ($self->full_path, @params);
    my $cmd_href = \@cmd;

    return $cmd_href;
}

# Maybe this is not very clean, but it works (except on MS Windows)
sub ack {
    my ($self, $cmd) = @_;
    my (
	$success,
	$error_message,
	$full_buffer,
	$stdout_buffer,
	$stderr_buffer
    ) = run (
	command => $cmd,
	verbose => 0
    );

    my @modules;

    debugf("All modules in distribution : " . join "", @$full_buffer);

    if ($success) {
	push @modules, split(/\n/m, $$full_buffer[0]);
    } else {
	say "No module have been found or IPC::Cmd failed with error $error_message";
    }

    return @modules;
}

1;

__END__

=pod

=head1 NAME

Dependencies::Searcher::AckRequester - Helps Dependencies::Searcher to use Ack

=cut

=head1 SYNOPSIS

    my $requester = Dependencies::Searcher::AckRequester->new();

    # Places to search...
    my @path = ("./lib", "./Makefile.PL", "./script");

    # Params for Ack
    my @params = ('--perl', '-hi', $pattern, @path);

    # Absolute path to the Ack binary
    my $ack_path = $requester->get_path();

    # Build the command for IPC::Cmd
    my $cmd_use = $requester->build_cmd(@params);

    # Execute the command and retrieve the output
    my @moduls = $requester->ack($cmd_use);

=head1 DESCRIPTION

This module use L<ack> through a system command to search recursively for
patterns. It use L<IPC::Cmd> as a layer between the module and L<ack>, that
execute and retrieve the command output.

It also builds the command itself (path and arguments). Arguments are
stored into an array, because it is too much dangerous to build a
command with strings (space problems are one reason among others).

It's not made to be used independantly from Dependencies::Searcher
(mean it's not supposed to be used directly into your programs, but
you can try if you want...)

=head1 SUBROUTINES/METHODS

=head2 get_path()

Returns the L<ack> full path if installed. Set the C<full_path>
L<Moose> attribute that will be used by ICP::Cmd. It verify also that
L<Ack> is reachable or warns about it.

=cut

=head2 build_cmd(@params)

C<build_cmd()> takes as parameter all the arguments Ack will
need. L<Dependencies::Searcher> defines it like this :

=over 4

=item * C<--perl>   : tells to search in Perl like files (C<*.pm>, C<*.pl>, etc.) 

=item * C<-hi>      : suppress the prefixing filename on output + ignore
case

=item * C<$pattern> : must be passed from your implementation

=item * C<@path>    : files and directories where L<ack> will go 

All these params are merged in an only array ref that is returned for
later use with L<IPC::Cmd>.

=back

=cut

=head2 ack($params_array_ref)

Execute the L<IPC::Cmd> command that calls C<ack> and returns an array of
potentially interesting lines, containing dependencies names but some
crap inside too.

=cut

=head1 CAVEATS

Win32 and Cygwin platforms aren't supported.

=head1 BUGS

Please report any bugs or feature requests to
C<bug-dependencies-searcher  at rt.cpan.org>, or through the web
interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Dependencies-Searcher>.
I will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 TODOs

=head1 AUTHOR

smonff, C<< <smonff at gmail.com> >>

=head1 ACKNOWLEDGEMENTS

=over

=item * Andy Lester's Ack

Ack gives me the wish to try to write this module. It was pure Perl so
I've choose it because it was possible to install it through CPAN
during the distribution installation process. Even if Ack was not
meant for being used programatically, this hacked use of Ack do the
job.

See L<http://beyondgrep.com/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2013 smonff.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut




Makefile.PL
use 5.010;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Dependencies::Searcher',
    AUTHOR           => q{smonff <smonff@gmail.com>},
    VERSION_FROM     => 'lib/Dependencies/Searcher.pm',
    ABSTRACT_FROM    => 'lib/Dependencies/Searcher.pm',
    LICENSE          => 'Perl_5',
    PL_FILES         => {},
    CONFIGURE_REQUIRES => {
	'ExtUtils::MakeMaker' => 6.74,
    },
    BUILD_REQUIRES => {
	'Test::More' => 0,
	'IO::File'   => 0,
    },
    TEST_REQUIRES => {
	'App::Ack'            => 2.10,
	'Module::Version'     => 0,
	'Test::Pod::Coverage' => 1.08,
	'Test::Pod'           => 1.48,
    },
    # For any module added, please add it to t/09-cpanfile.t too or
    # tests will fail :D
    PREREQ_PM => {
	'App::Ack'              => 2.10,
	'Data::Printer'         => 0.35,
	'File::HomeDir'         => 1.00,
	'File::Spec::Functions' => 3.40,
	'File::Stamped'         => 0.03,
	'IPC::Cmd'              => 0,
	'IPC::Run'              => 0.90,
	'Log::Minimal'          => 0.16,
	'Module::CoreList'      => 2.99,
	'Module::Version'       => 0.12,
	'Moose'                 => 2.0602,
	'Path::Class'           => 0.32,
	'Test::More'            => 0,
	'Version::Compare'      => 0.14,
    },

    MIN_PERL_VERSION => '5.010',
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Dependencies-Searcher-*' },
);
